File: problems/32_pandigital_products/python/pandigital_products.py
# pandigital_products.py
# Project Euler 32 - Pandigital Products (Python)

def is_pandigital_identity(a: int, b: int) -> bool:
    p = a * b
    s = f"{a}{b}{p}"
    if len(s) != 9:
        return False
    seen = [False] * 10
    for ch in s:
        if ch == '0':
            return False
        d = ord(ch) - ord('0')
        if seen[d]:
            return False
        seen[d] = True
    return all(seen[d] for d in range(1, 10))

def main() -> None:
    products = set()

    for a in range(1, 10):
        for b in range(1000, 10000):
            if is_pandigital_identity(a, b):
                products.add(a * b)

    for a in range(10, 100):
        for b in range(100, 1000):
            if is_pandigital_identity(a, b):
                products.add(a * b)

    total = sum(products)
    # print(total)  # TODO: reveal locally
    print("sum = <redacted>")

if __name__ == "__main__":
    main()

File: problems/32_pandigital_products/go/pandigital_products.go
// pandigital_products.go
// Project Euler 32 - Pandigital Products (Go)
package main

import "fmt"

func isPandigital(a, b int) bool {
    p := a * b
    s := fmt.Sprintf("%d%d%d", a, b, p)
    if len(s) != 9 {
        return false
    }
    seen := [10]bool{}
    for i := 0; i < len(s); i++ {
        d := int(s[i] - '0')
        if d == 0 || seen[d] {
            return false
        }
        seen[d] = true
    }
    for d := 1; d <= 9; d++ {
        if !seen[d] {
            return false
        }
    }
    return true
}

func main() {
    products := map[int]struct{}{}

    for a := 1; a <= 9; a++ {
        for b := 1000; b <= 9999; b++ {
            if isPandigital(a, b) {
                products[a*b] = struct{}{}
            }
        }
    }

    for a := 10; a <= 99; a++ {
        for b := 100; b <= 999; b++ {
            if isPandigital(a, b) {
                products[a*b] = struct{}{}
            }
        }
    }

    sum := 0
    for p := range products {
        sum += p
    }
    // fmt.Println(sum) // TODO: reveal locally
    fmt.Println("sum = <redacted>")
}
